hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/RepeatedOmKeyInfo.java

```diff
@@ -66,7 +68,7 @@ public static RepeatedOmKeyInfo getFromProto(RepeatedKeyInfo
    * @return
    */
   public RepeatedKeyInfo getProto(boolean compact, int clientVersion) {
-    List<KeyInfo> list = new ArrayList<>();
+    List<KeyInfo> list = new CopyOnWriteArrayList<>();
     for(OmKeyInfo k : omKeyInfoList) {
```

Comment:
Since the concurrent modification is reported during iteration, I think the problem is with `omKeyInfoList`, not `list`.

Reply:
Thanks @adoroszlai  can review this.
This exception is 100% reproducible when we use FIO concurrent testing.  Looking at the stack, we can see that it was a thread safety problem of RepeatedOmKeyInfo, and the problem did not reappear after this fix.

I checked that a similar error had occurred in OM before, when TreeMap was used in OmMultipartKeyInfo：https://issues.apache.org/jira/browse/HDDS-2356

Recreating this problem with UT can be complex, and currently we only encounter this problem with FUSE and FIO concurrent writes. We have not found this problem in other operations.

Reply:
Repro:

```java
  @Test
  public void threadSafety() throws InterruptedException {
    final OmKeyInfo key = getKeyInfo(1);
    final RepeatedOmKeyInfo subject = new RepeatedOmKeyInfo(key);
    final RepeatedOmKeyInfoCodec codec = new RepeatedOmKeyInfoCodec(true);
    final AtomicBoolean failed = new AtomicBoolean();
    ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true)
        .build();
    threadFactory.newThread(() -> {
      for (int i = 0; i < 1000000; i++) {
        try {
          codec.toPersistedFormat(subject);
        } catch (Exception e) {
          e.printStackTrace();
          failed.set(true);
        }
      }
    }).start();
    threadFactory.newThread(() -> {
      for (int i = 0; i < 10000; i++) {
        subject.addOmKeyInfo(key);
      }
    }).start();
    final long start = Time.monotonicNow();
    while (!failed.get() && (Time.monotonicNow() - start < 5000)) {
      Thread.sleep(100);
    }
    assertFalse(failed.get());
  }
```

```
java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
	at java.util.ArrayList$Itr.next(ArrayList.java:859)
	at org.apache.hadoop.ozone.om.helpers.RepeatedOmKeyInfo.getProto(RepeatedOmKeyInfo.java:70)
	at org.apache.hadoop.ozone.om.codec.RepeatedOmKeyInfoCodec.toPersistedFormat(RepeatedOmKeyInfoCodec.java:50)
```

Reply:
I think the correct way to fix is implement copyObject in RepeatedOmKeyInfo properly right now it is returning same object. I believe this might be exposed due to recent fixes, where we are adding RepeatedOmKeyInfo to cache to avoid cleanup when override.

For example refer similar thing is observed in OMKeyInfo. 
https://issues.apache.org/jira/browse/HDDS-2322

Some background when we get from Table#get API's if we get new object, then there should be no race conditions or multiple threads modifiying same object and we will not see this issue.
You can refer this where we introduced CopyObject to solve this kind of issue.

https://issues.apache.org/jira/browse/HDDS-2344

Reply:
Thanks @adoroszlai and @bharatviswa504 for you comment. I will modify this PR as suggested.